src/llamafactory/data/mm_plugin.py

```diff
@@ -447,6 +447,80 @@ def get_mm_inputs(
         mm_inputs["token_type_ids"] = _get_paligemma_token_type_ids(imglens, seqlens, processor)
         return mm_inputs
 
+class PixtralPlugin(BasePlugin):
+    @override
+    def process_messages(
+        self,
+        messages: Sequence[Dict[str, str]],
+        images: Sequence["ImageInput"],
+        videos: Sequence["VideoInput"],
+        processor: Optional["ProcessorMixin"],
+    ) -> List[Dict[str, str]]:
+        patch_size = processor.patch_size
+        image_token = processor.image_token
+        image_break_token = processor.image_break_token
+        image_end_token = processor.image_end_token
+
+        self._validate_input(images, videos)
+        num_image_tokens = 0
+        img_kwargs = self._get_mm_inputs(images, videos, processor)
+        image_input_sizes = None
+
+        if img_kwargs.get("pixel_values") is not None:
+            image_input_sizes = img_kwargs["image_sizes"]
+
+        messages = deepcopy(messages)
+        for message in messages:
+            content = message["content"]
+            img_id = 0
+            while IMAGE_PLACEHOLDER in content:
+
+                if image_input_sizes is None:
+                    raise ValueError("The number of images does not match the number of {} tokens".format(IMAGE_PLACEHOLDER))
+
+                image_size =  image_input_sizes[0][img_id]
+                height, width = image_size
+                num_height_tokens = height // patch_size
+                num_width_tokens = width // patch_size
+                replace_tokens = [
+                    [image_token] * num_width_tokens + [image_break_token]
+                ] * num_height_tokens
+                # Flatten list
+                replace_tokens = [item for sublist in replace_tokens for item in sublist]
+                replace_tokens[-1] = image_end_token
+                replace_str = "".join(replace_tokens)
+                content = content.replace(IMAGE_PLACEHOLDER, replace_str, 1)
+
+                img_id += 1
+                num_image_tokens += 1
+
+            message["content"] = content
+
+        if len(images) != num_image_tokens:
+            raise ValueError("The number of images does not match the number of {} tokens".format(IMAGE_PLACEHOLDER))
+
+        return messages
+
+    @override
+    def get_mm_inputs(
+        self,
+        images: Sequence["ImageInput"],
+        videos: Sequence["VideoInput"],
+        imglens: Sequence[int],
+        vidlens: Sequence[int],
+        seqlens: Sequence[int],
+        processor: Optional["ProcessorMixin"],
+    ) -> Dict[str, Union[List[int], "torch.Tensor"]]:
+        self._validate_input(images, videos)
+        mm_inputs = self._get_mm_inputs(images, videos, processor)
+        # hack for hf engine
+        if mm_inputs.get("pixel_values") and len(mm_inputs.get("pixel_values")[0]) == 1:
+            mm_inputs["pixel_values"] = mm_inputs["pixel_values"][0][0].unsqueeze(0)
```

Comment:
Please give me an explanation

Reply:
I realize this part of the code appears somewhat weird.

The `pixtral.image_processor` takes `images_list` as input and outputs a dictionary in the format `{"pixel_values": images_list, "image_sizes": batch_image_sizes}`. The `pixel_values` is a list causing errors when passed to `hf_engine`. To address this, I attempted to extract the vectors in the form `[c, h, w]` and reshape them into `[b, c, w, h]` as a simple way to adapt to `hf_engine`'s input requirements. 

---

However, another issue arises during training when `batchsize > 2`. When a batch contains images of varying sizes, they cannot be stacked directly into `[b, c, w, h]` tensors using `torch.stack`. I tried resizing all the images to a fixed size and then wrapping them into `[b, c, w, h]` tensors, but during testing, I found that this approach negatively impacted model performance. I suspect this might be because the resized input images are OOD compared to the pretraining images.

---

**But I'm sorry for not fully validating the training process when `bs > 2` and there are still some issues.** 

Currently, during training with `bs > 2`, the forward process results in a mismatch between the number of images (which becomes 1) and the batch size. 

I will continue working on this issue. 

---
A comparison of the resize effects is shown in the figures below:

original:
![3013450a845c907d7a8c4360f402aaa](https://github.com/user-attachments/assets/fd7dadfd-28b0-4ebc-bf13-86e244ab085f)

resized image to (512, 512):
![5420a4a97ba2d09411313ea3bb00a7c](https://github.com/user-attachments/assets/06072750-49e8-415a-9c69-2f54a7cc962a)

---
Now `bs=1` per device works well in training.
![eff](https://github.com/user-attachments/assets/3bc96d93-53e5-4226-8115-9cd980d7378a)